File: 39.组合总和.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start

class Solution
{
public:
    vector<vector<int>> result;
    vector<int> current;
    int sum;
    void findVector(vector<int> &candidates, int target, int startIndex)
    {
        if (sum > target)
        {
            return;
        }
        if (sum == target)
        {
            result.push_back(current);
            return;
        }

        for (int i = startIndex; i < candidates.size(); i++)
        {
            current.push_back(candidates[i]);
            sum += candidates[i];
            findVector(candidates, target, i);
            current.pop_back();
            sum -= candidates[i];
        }
    }
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {

        findVector(candidates, target, 0);
        return result;
    }
};
// @lc code=end

File: Back Tracking/求和/39.组合总和.ts
/*
 * @lc app=leetcode.cn id=39 lang=typescript
 *
 * [39] 组合总和
 */

// @lc code=start
function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = []

  const backTracking = (arr: number[], sum: number, start: number) => {
    if (sum === target) {
      result.push([...arr])
      return
    }

    // 排序后才有使用此处的剪枝操作，越往后数字越大，所以不需要继续加了
    if (sum > target) {
      return
    }

    for (let i = start; i < candidates.length; i++) {
      arr.push(candidates[i])
      // 元素可以重复，所以此处是i而不是i+1
      backTracking(arr, sum + candidates[i], i)
      arr.pop()
    }
  }
  // 对数组升序排列，方便下一步进行剪枝操作，注意，升序之后才能剪枝
  candidates = candidates.sort((a, b) => a - b)
  backTracking([], 0, 0)

  return result
};
// @lc code=end

